api/proto/teleport/legacy/types/types.proto

```diff
@@ -3276,14 +3276,8 @@ message LoginStatus {
     (gogoproto.nullable) = false,
     (gogoproto.jsontag) = "lock_expires,omitempty"
   ];
-  // RecoveryAttemptLockExpires contains the time when this lock will expire
-  // from reaching MaxAccountRecoveryAttempts. This field is used to determine
-  // if a user got locked from recovery attempts.
-  google.protobuf.Timestamp RecoveryAttemptLockExpires = 5 [
-    (gogoproto.stdtime) = true,
-    (gogoproto.nullable) = false,
-    (gogoproto.jsontag) = "recovery_attempt_lock_expires,omitempty"
-  ];
+  reserved 5; // removed "google.protobuf.Timestamp RecoveryAttemptLockExpires" after lockout was removed
+  reserved "RecoveryAttemptLockExpires";
```

Comment:
The removal of this field is making [UpdateAndSwapUser][1] fail.

If the stored user had the "recovery_attempt_lock_expires" field set, then the comparison always fails. That is because when we read and unmarshal the user [the unknown field gets dropped][2], but the actual CompareAndSwap operation compares [using the storage blobs instead][3].

Reverting the deletion, but keeping the field deprecated should serve as an immediate fix. Arguably, a better fix is to make sure CompareAndSwap compares apples-to-apples: either user vs user, or blob vs blob, but never user vs blob.

@jentfoo @rosstimothy

[1]: https://github.com/gravitational/teleport/blob/b8deacfaffc25073f22ff0dadf157344b2fd2688/lib/services/local/users.go#L383

[2]: https://github.com/gravitational/teleport/blob/b8deacfaffc25073f22ff0dadf157344b2fd2688/lib/services/local/users.go#L384

[3]: https://github.com/gravitational/teleport/blob/b8deacfaffc25073f22ff0dadf157344b2fd2688/lib/backend/lite/lite.go#L461

Reply:
I just submitted a PR to bring the field back: https://github.com/gravitational/teleport/pull/37618

Thank you for finding this issue @codingllama!